### PR TITLE
[FLINK-14789] [table-planner-blink] extends max/min type in ColumnStats from Number to Comparable

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*extends max/min type in ColumnStats from Number to Comparable the help planner to find a better plan if a query has predicate on non-number column(e.g. date, time, timestamp)*

## Brief change log

  - *change max/min type from Number to Comparable in ColumnStats*
  - *convert max/min info in CatalogColumnStatisticsDataDate to ColumnStats*

## Verifying this change

This change added tests and can be verified as follows:

  - *Updated SelectivityEstimatorTest for selectivity on Boolean, Date, Time, Timestamp, String column*
  - *Updated CatalogStatisticsTest for date max/min*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java [hunk 1]
```diff
@@ -50,22 +50,22 @@
...
+	private final Comparable<?> min;
 
 	public ColumnStats(
 		Long ndv,
 		Long nullCount,
 		Double avgLen,
 		Integer maxLen,
-		Number max,
-		Number min) {
+		Comparable<?> max,
```
Comment (KurtYoung) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java]: This is API breaking change, right?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java [hunk 2]
```diff
@@ -50,28 +50,72 @@
...
+	 * Deprecated because not well supported comparable type,
+	 * e.g. {@link java.util.Date}, {@link java.sql.Timestamp}.
 	 */
-	private final Number max;
+	@Deprecated
+	private final Number maxValue;
 
 	/**
-	 * min value of column values.
+	 * max value of column values, null if the value is not comparable.
```
Comment (KurtYoung) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java]: `null if the value is not comparable`? 
I thought null means we don't have such information.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java [hunk 3]
```diff
@@ -115,20 +189,81 @@ public String toString() {
 		if (max != null) {
 			columnStats.add("max=" + max);
 		}
+		if (maxValue != null) {
```
Comment (KurtYoung) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java]: we should only output one `max`, and the old way should have higher priority.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java [hunk 4]
```diff
@@ -115,20 +189,81 @@ public String toString() {
...
 
 	/**
 	 * Create a deep copy of "this" instance.
+	 *
 	 * @return a deep copy
 	 */
 	public ColumnStats copy() {
-		return new ColumnStats(this.ndv, this.nullCount, this.avgLen, this.maxLen,
-			this.max, this.min);
+		if (maxValue != null) {
```
Comment (KurtYoung) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java]: what if user didn't set `maxValue` but set `minValue`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnInterval.scala [hunk 5]
```diff
@@ -66,12 +66,20 @@ class FlinkRelMdColumnInterval private extends MetadataHandler[ColumnInterval] {
...
-      if (min == null && max == null) {
-        null
-      } else {
+      val minValue = colStats.getMinValue
+      val maxValue = colStats.getMaxValue
+      val min = colStats.getMin
+      val max = colStats.getMax
+
+      Preconditions.checkArgument(
+        (minValue == null && maxValue == null) || (max == null && min == null))
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnInterval.scala]: Why checking this? I think it's valid to set only max value or min value.

**Comment by KurtYoung** [comment]
cc @dawidwys @twalthr

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java (hunk 3 above)
Comment (godfreyhe) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java]: `max` and `maxValue` could not be null both, so only one `max` will print if the value is not null

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java (hunk 4 above)
Comment (godfreyhe) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/plan/stats/ColumnStats.java]: `minValue != null` missing, full condition should be `maxValue != null || minValue != null`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnInterval.scala (hunk 5 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdColumnInterval.scala]: the validation makes sure the values of `minValue/maxValue` and `min/max` could be not null both.

**Comment by godfreyhe** [comment]
@KurtYoung thanks for the suggestion, i have updated the pr

**Pull request merged**